{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "pharmaceutical-notification",
   "metadata": {},
   "source": [
    "# Identify bacterial genera associated with metabolites\n",
    "\n",
    "We will now select the microbes significantly associated with each metabolite. We will use the genus level here since that is usually the lowest taxonomic rank we can confidently identify with 16S amplicon sequencing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "lyric-christmas",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "410"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "train = pd.read_csv(\"data/train.csv\")\n",
    "valid = pd.read_csv(\"data/valid.csv\")\n",
    "abundance = pd.read_csv(\"/proj/arivale/microbiome/16S_processed/asvs.csv\")\n",
    "abundance[\"stool_sample_id\"] = abundance.id.str.split(\"|\").str[0]\n",
    "abundance = abundance[abundance.stool_sample_id.isin(train.stool_sample_id) | abundance.stool_sample_id.isin(valid.stool_sample_id)]\n",
    "\n",
    "taxonomy = pd.read_csv(\"/proj/arivale/microbiome/16S_processed/taxonomy.csv\")\n",
    "taxa_compact = taxonomy.Family + \"|\" + taxonomy.Genus\n",
    "taxa_compact.index = taxonomy.id\n",
    "abundance[\"taxon\"] = taxa_compact[abundance.hash].values\n",
    "genera = abundance.groupby([\"stool_sample_id\", \"taxon\"]).sum().reset_index()\n",
    "genera.taxon.nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "further-scroll",
   "metadata": {},
   "source": [
    "We also filter out genera that were not observed in at least 50% of all individuals and that have less than 10 detected reads per sample on average. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "current-protocol",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "82"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prevalence = genera.groupby(\"taxon\").apply(lambda df: sum(df[\"count\"] > 0)) / genera.stool_sample_id.nunique()\n",
    "means = genera.groupby(\"taxon\")[\"count\"].mean()\n",
    "good = (prevalence > 0.5) & (means > 10)\n",
    "genera_filtered = genera[genera.taxon.isin(good.index[good])]\n",
    "genera_filtered.taxon.nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "difficult-ambassador",
   "metadata": {},
   "source": [
    "Now we will CLR transform the abundances."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "prospective-plaintiff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1569, 82)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "def clr(x, pseudo):\n",
    "    if pseudo == 0:\n",
    "        out = x.copy()\n",
    "        out[x == 0] = float(\"nan\")\n",
    "        out[x > 0] = np.log(x[x > 0]) - np.mean(np.log(x[x > 0])) \n",
    "    else:\n",
    "        out = np.log(x + pseudo) - np.mean(np.log(x + pseudo))\n",
    "    return out\n",
    "\n",
    "genera_wide = pd.pivot_table(genera_filtered, index=\"stool_sample_id\", columns=\"taxon\", values=\"count\", fill_value=0)\n",
    "genera_wide_clr = genera_wide.apply(lambda x: clr(x, 1), axis=1)\n",
    "genera_wide_clr.to_csv(\"data/genera_clr_filtered.csv\")\n",
    "\n",
    "genera_wide_clr_drop = genera_wide.apply(lambda x: clr(x, 0), axis=1)\n",
    "genera_wide_clr_drop.to_csv(\"data/genera_clr_filtered_nonzero.csv\")\n",
    "genera_wide_clr.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "forward-recruitment",
   "metadata": {},
   "outputs": [],
   "source": [
    "metabolites_and_microbes = pd.merge(train, genera_wide_clr, left_on=\"stool_sample_id\", right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "wireless-updating",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>public_client_id</th>\n",
       "      <th>genome_id</th>\n",
       "      <th>blood_sample_id</th>\n",
       "      <th>blood_days_in_program</th>\n",
       "      <th>stool_sample_id</th>\n",
       "      <th>stool_days_in_program</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>stool_vendor</th>\n",
       "      <th>WEIGHT_CALC</th>\n",
       "      <th>...</th>\n",
       "      <th>Ruminococcaceae|Ruminococcaceae_UCG-014</th>\n",
       "      <th>Ruminococcaceae|Ruminococcus_1</th>\n",
       "      <th>Ruminococcaceae|Ruminococcus_2</th>\n",
       "      <th>Ruminococcaceae|Subdoligranulum</th>\n",
       "      <th>Ruminococcaceae|UBA1819</th>\n",
       "      <th>Streptococcaceae|Lactococcus</th>\n",
       "      <th>Streptococcaceae|Streptococcus</th>\n",
       "      <th>Tannerellaceae|Parabacteroides</th>\n",
       "      <th>Veillonellaceae|Dialister</th>\n",
       "      <th>Veillonellaceae|Veillonella</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1241633</td>\n",
       "      <td>A789AU958-007</td>\n",
       "      <td>A789AU992-002</td>\n",
       "      <td>36.0</td>\n",
       "      <td>AV15-1655</td>\n",
       "      <td>36.0</td>\n",
       "      <td>M</td>\n",
       "      <td>71.0</td>\n",
       "      <td>SecondGenome</td>\n",
       "      <td>179.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.462865</td>\n",
       "      <td>3.690187</td>\n",
       "      <td>-0.061668</td>\n",
       "      <td>5.699964</td>\n",
       "      <td>1.513869</td>\n",
       "      <td>-1.265641</td>\n",
       "      <td>2.477306</td>\n",
       "      <td>1.506948</td>\n",
       "      <td>4.792963</td>\n",
       "      <td>-0.284811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1020175</td>\n",
       "      <td>DS-268800</td>\n",
       "      <td>A430BH742-003</td>\n",
       "      <td>8.0</td>\n",
       "      <td>22001612561075</td>\n",
       "      <td>13.0</td>\n",
       "      <td>F</td>\n",
       "      <td>41.0</td>\n",
       "      <td>DNAGenotek</td>\n",
       "      <td>192.8</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.151371</td>\n",
       "      <td>2.642095</td>\n",
       "      <td>-4.151371</td>\n",
       "      <td>2.343894</td>\n",
       "      <td>0.892054</td>\n",
       "      <td>-0.413702</td>\n",
       "      <td>0.825362</td>\n",
       "      <td>5.826435</td>\n",
       "      <td>3.427797</td>\n",
       "      <td>-1.015877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1230573</td>\n",
       "      <td>DS-284622</td>\n",
       "      <td>A773BJ600-002</td>\n",
       "      <td>4.0</td>\n",
       "      <td>22001701512104</td>\n",
       "      <td>22.0</td>\n",
       "      <td>F</td>\n",
       "      <td>60.0</td>\n",
       "      <td>DNAGenotek</td>\n",
       "      <td>152.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.840533</td>\n",
       "      <td>-4.836265</td>\n",
       "      <td>-4.836265</td>\n",
       "      <td>-4.836265</td>\n",
       "      <td>0.436734</td>\n",
       "      <td>-1.122693</td>\n",
       "      <td>-1.029603</td>\n",
       "      <td>3.860078</td>\n",
       "      <td>3.405438</td>\n",
       "      <td>-4.836265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1687802</td>\n",
       "      <td>A719AT743-004</td>\n",
       "      <td>A719AT785-002</td>\n",
       "      <td>88.0</td>\n",
       "      <td>AV15-1320</td>\n",
       "      <td>88.0</td>\n",
       "      <td>F</td>\n",
       "      <td>53.0</td>\n",
       "      <td>SecondGenome</td>\n",
       "      <td>242.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.409260</td>\n",
       "      <td>3.829955</td>\n",
       "      <td>3.229308</td>\n",
       "      <td>1.300270</td>\n",
       "      <td>0.598073</td>\n",
       "      <td>1.155088</td>\n",
       "      <td>2.365292</td>\n",
       "      <td>-3.409260</td>\n",
       "      <td>-3.409260</td>\n",
       "      <td>0.117101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1864964</td>\n",
       "      <td>DS-267221</td>\n",
       "      <td>A862BF047-008</td>\n",
       "      <td>14.0</td>\n",
       "      <td>22001612561843</td>\n",
       "      <td>15.0</td>\n",
       "      <td>F</td>\n",
       "      <td>43.0</td>\n",
       "      <td>DNAGenotek</td>\n",
       "      <td>170.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.535311</td>\n",
       "      <td>4.653378</td>\n",
       "      <td>1.226863</td>\n",
       "      <td>2.005952</td>\n",
       "      <td>-0.762722</td>\n",
       "      <td>-3.535311</td>\n",
       "      <td>0.942026</td>\n",
       "      <td>4.230682</td>\n",
       "      <td>-3.535311</td>\n",
       "      <td>-0.827261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>1035670</td>\n",
       "      <td>DS-269430</td>\n",
       "      <td>A808BG205-003</td>\n",
       "      <td>33.0</td>\n",
       "      <td>22001612562579</td>\n",
       "      <td>9.0</td>\n",
       "      <td>F</td>\n",
       "      <td>33.0</td>\n",
       "      <td>DNAGenotek</td>\n",
       "      <td>155.2</td>\n",
       "      <td>...</td>\n",
       "      <td>-5.483529</td>\n",
       "      <td>4.261138</td>\n",
       "      <td>3.655530</td>\n",
       "      <td>4.293317</td>\n",
       "      <td>-0.608332</td>\n",
       "      <td>-2.392487</td>\n",
       "      <td>0.461891</td>\n",
       "      <td>1.905417</td>\n",
       "      <td>2.450267</td>\n",
       "      <td>-2.187693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>1013950</td>\n",
       "      <td>A752AT594-004</td>\n",
       "      <td>A752AT607-002</td>\n",
       "      <td>77.0</td>\n",
       "      <td>AV15-1319</td>\n",
       "      <td>77.0</td>\n",
       "      <td>F</td>\n",
       "      <td>42.0</td>\n",
       "      <td>SecondGenome</td>\n",
       "      <td>126.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.016214</td>\n",
       "      <td>2.118971</td>\n",
       "      <td>4.178741</td>\n",
       "      <td>3.622815</td>\n",
       "      <td>0.170962</td>\n",
       "      <td>-5.147158</td>\n",
       "      <td>2.745668</td>\n",
       "      <td>1.831056</td>\n",
       "      <td>-5.147158</td>\n",
       "      <td>-5.147158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>1244110</td>\n",
       "      <td>A939AV031-007</td>\n",
       "      <td>A473AS982-002</td>\n",
       "      <td>52.0</td>\n",
       "      <td>AV15-1143</td>\n",
       "      <td>52.0</td>\n",
       "      <td>F</td>\n",
       "      <td>62.0</td>\n",
       "      <td>SecondGenome</td>\n",
       "      <td>126.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.807030</td>\n",
       "      <td>1.858001</td>\n",
       "      <td>2.510743</td>\n",
       "      <td>1.855598</td>\n",
       "      <td>-2.277166</td>\n",
       "      <td>-1.717550</td>\n",
       "      <td>1.581457</td>\n",
       "      <td>1.387677</td>\n",
       "      <td>2.798008</td>\n",
       "      <td>-5.272898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>1329753</td>\n",
       "      <td>DS-265707</td>\n",
       "      <td>A579BF997-003</td>\n",
       "      <td>3.0</td>\n",
       "      <td>22001612560465</td>\n",
       "      <td>15.0</td>\n",
       "      <td>M</td>\n",
       "      <td>40.0</td>\n",
       "      <td>DNAGenotek</td>\n",
       "      <td>269.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.352577</td>\n",
       "      <td>3.215964</td>\n",
       "      <td>3.505026</td>\n",
       "      <td>3.807427</td>\n",
       "      <td>-3.081145</td>\n",
       "      <td>-3.081145</td>\n",
       "      <td>0.870099</td>\n",
       "      <td>1.429714</td>\n",
       "      <td>-3.081145</td>\n",
       "      <td>-3.081145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>1028150</td>\n",
       "      <td>DS-268107</td>\n",
       "      <td>A202BI892-020</td>\n",
       "      <td>30.0</td>\n",
       "      <td>22001612562223</td>\n",
       "      <td>25.0</td>\n",
       "      <td>M</td>\n",
       "      <td>57.0</td>\n",
       "      <td>DNAGenotek</td>\n",
       "      <td>225.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.794583</td>\n",
       "      <td>4.623733</td>\n",
       "      <td>3.963858</td>\n",
       "      <td>-0.579437</td>\n",
       "      <td>0.609202</td>\n",
       "      <td>-0.051369</td>\n",
       "      <td>1.912690</td>\n",
       "      <td>3.838356</td>\n",
       "      <td>3.229638</td>\n",
       "      <td>-4.408078</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows Ã— 1030 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     public_client_id      genome_id blood_sample_id  blood_days_in_program  \\\n",
       "0             1241633  A789AU958-007   A789AU992-002                   36.0   \n",
       "1             1020175      DS-268800   A430BH742-003                    8.0   \n",
       "2             1230573      DS-284622   A773BJ600-002                    4.0   \n",
       "3             1687802  A719AT743-004   A719AT785-002                   88.0   \n",
       "4             1864964      DS-267221   A862BF047-008                   14.0   \n",
       "..                ...            ...             ...                    ...   \n",
       "995           1035670      DS-269430   A808BG205-003                   33.0   \n",
       "996           1013950  A752AT594-004   A752AT607-002                   77.0   \n",
       "997           1244110  A939AV031-007   A473AS982-002                   52.0   \n",
       "998           1329753      DS-265707   A579BF997-003                    3.0   \n",
       "999           1028150      DS-268107   A202BI892-020                   30.0   \n",
       "\n",
       "    stool_sample_id  stool_days_in_program sex   age  stool_vendor  \\\n",
       "0         AV15-1655                   36.0   M  71.0  SecondGenome   \n",
       "1    22001612561075                   13.0   F  41.0    DNAGenotek   \n",
       "2    22001701512104                   22.0   F  60.0    DNAGenotek   \n",
       "3         AV15-1320                   88.0   F  53.0  SecondGenome   \n",
       "4    22001612561843                   15.0   F  43.0    DNAGenotek   \n",
       "..              ...                    ...  ..   ...           ...   \n",
       "995  22001612562579                    9.0   F  33.0    DNAGenotek   \n",
       "996       AV15-1319                   77.0   F  42.0  SecondGenome   \n",
       "997       AV15-1143                   52.0   F  62.0  SecondGenome   \n",
       "998  22001612560465                   15.0   M  40.0    DNAGenotek   \n",
       "999  22001612562223                   25.0   M  57.0    DNAGenotek   \n",
       "\n",
       "     WEIGHT_CALC  ...  Ruminococcaceae|Ruminococcaceae_UCG-014  \\\n",
       "0          179.0  ...                                -3.462865   \n",
       "1          192.8  ...                                -4.151371   \n",
       "2          152.0  ...                                -1.840533   \n",
       "3          242.0  ...                                -3.409260   \n",
       "4          170.0  ...                                -3.535311   \n",
       "..           ...  ...                                      ...   \n",
       "995        155.2  ...                                -5.483529   \n",
       "996        126.0  ...                                 3.016214   \n",
       "997        126.0  ...                                 2.807030   \n",
       "998        269.0  ...                                 2.352577   \n",
       "999        225.0  ...                                 2.794583   \n",
       "\n",
       "    Ruminococcaceae|Ruminococcus_1 Ruminococcaceae|Ruminococcus_2  \\\n",
       "0                         3.690187                      -0.061668   \n",
       "1                         2.642095                      -4.151371   \n",
       "2                        -4.836265                      -4.836265   \n",
       "3                         3.829955                       3.229308   \n",
       "4                         4.653378                       1.226863   \n",
       "..                             ...                            ...   \n",
       "995                       4.261138                       3.655530   \n",
       "996                       2.118971                       4.178741   \n",
       "997                       1.858001                       2.510743   \n",
       "998                       3.215964                       3.505026   \n",
       "999                       4.623733                       3.963858   \n",
       "\n",
       "     Ruminococcaceae|Subdoligranulum  Ruminococcaceae|UBA1819  \\\n",
       "0                           5.699964                 1.513869   \n",
       "1                           2.343894                 0.892054   \n",
       "2                          -4.836265                 0.436734   \n",
       "3                           1.300270                 0.598073   \n",
       "4                           2.005952                -0.762722   \n",
       "..                               ...                      ...   \n",
       "995                         4.293317                -0.608332   \n",
       "996                         3.622815                 0.170962   \n",
       "997                         1.855598                -2.277166   \n",
       "998                         3.807427                -3.081145   \n",
       "999                        -0.579437                 0.609202   \n",
       "\n",
       "     Streptococcaceae|Lactococcus  Streptococcaceae|Streptococcus  \\\n",
       "0                       -1.265641                        2.477306   \n",
       "1                       -0.413702                        0.825362   \n",
       "2                       -1.122693                       -1.029603   \n",
       "3                        1.155088                        2.365292   \n",
       "4                       -3.535311                        0.942026   \n",
       "..                            ...                             ...   \n",
       "995                     -2.392487                        0.461891   \n",
       "996                     -5.147158                        2.745668   \n",
       "997                     -1.717550                        1.581457   \n",
       "998                     -3.081145                        0.870099   \n",
       "999                     -0.051369                        1.912690   \n",
       "\n",
       "     Tannerellaceae|Parabacteroides  Veillonellaceae|Dialister  \\\n",
       "0                          1.506948                   4.792963   \n",
       "1                          5.826435                   3.427797   \n",
       "2                          3.860078                   3.405438   \n",
       "3                         -3.409260                  -3.409260   \n",
       "4                          4.230682                  -3.535311   \n",
       "..                              ...                        ...   \n",
       "995                        1.905417                   2.450267   \n",
       "996                        1.831056                  -5.147158   \n",
       "997                        1.387677                   2.798008   \n",
       "998                        1.429714                  -3.081145   \n",
       "999                        3.838356                   3.229638   \n",
       "\n",
       "     Veillonellaceae|Veillonella  \n",
       "0                      -0.284811  \n",
       "1                      -1.015877  \n",
       "2                      -4.836265  \n",
       "3                       0.117101  \n",
       "4                      -0.827261  \n",
       "..                           ...  \n",
       "995                    -2.187693  \n",
       "996                    -5.147158  \n",
       "997                    -5.272898  \n",
       "998                    -3.081145  \n",
       "999                    -4.408078  \n",
       "\n",
       "[1000 rows x 1030 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabolites_and_microbes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "following-estate",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "54c4393eddf644f3ba85484b8f60aa9c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-6ef812f28ac2>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0mformula\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"{} ~ Q('{}')\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mcomb\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mtrack\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcombinations\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 13\u001b[0;31m     \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmetabolites_and_microbes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcomb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdropna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     14\u001b[0m     \u001b[0mptest\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpearsonr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m     \u001b[0mmetrics\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcomb\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"r\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mptest\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/gxe2021/lib/python3.8/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3034\u001b[0m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwhere\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3035\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3036\u001b[0;31m         \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_take_with_is_copy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3037\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3038\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mis_single_key\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/gxe2021/lib/python3.8/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_take_with_is_copy\u001b[0;34m(self, indices, axis)\u001b[0m\n\u001b[1;32m   3598\u001b[0m         \u001b[0mSee\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mdocstring\u001b[0m \u001b[0mof\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m`\u001b[0m\u001b[0mtake\u001b[0m\u001b[0;31m`\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mfull\u001b[0m \u001b[0mexplanation\u001b[0m \u001b[0mof\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mparameters\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3599\u001b[0m         \"\"\"\n\u001b[0;32m-> 3600\u001b[0;31m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindices\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mindices\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3601\u001b[0m         \u001b[0;31m# Maybe set copy if we didn't actually change the index.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3602\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mequals\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/gxe2021/lib/python3.8/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mtake\u001b[0;34m(self, indices, axis, is_copy, **kwargs)\u001b[0m\n\u001b[1;32m   3584\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_consolidate_inplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3585\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3586\u001b[0;31m         new_data = self._mgr.take(\n\u001b[0m\u001b[1;32m   3587\u001b[0m             \u001b[0mindices\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_block_manager_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverify\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3588\u001b[0m         )\n",
      "\u001b[0;32m~/miniconda3/envs/gxe2021/lib/python3.8/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mtake\u001b[0;34m(self, indexer, axis, verify, convert)\u001b[0m\n\u001b[1;32m   1472\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1473\u001b[0m         \u001b[0mnew_labels\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maxes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1474\u001b[0;31m         return self.reindex_indexer(\n\u001b[0m\u001b[1;32m   1475\u001b[0m             \u001b[0mnew_axis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnew_labels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mallow_dups\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1476\u001b[0m         )\n",
      "\u001b[0;32m~/miniconda3/envs/gxe2021/lib/python3.8/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mreindex_indexer\u001b[0;34m(self, new_axis, indexer, axis, fill_value, allow_dups, copy, consolidate, only_slice)\u001b[0m\n\u001b[1;32m   1305\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1306\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0maxis\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1307\u001b[0;31m             new_blocks = self._slice_take_blocks_ax0(\n\u001b[0m\u001b[1;32m   1308\u001b[0m                 \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfill_value\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfill_value\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0monly_slice\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0monly_slice\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1309\u001b[0m             )\n",
      "\u001b[0;32m~/miniconda3/envs/gxe2021/lib/python3.8/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36m_slice_take_blocks_ax0\u001b[0;34m(self, slice_or_indexer, fill_value, only_slice)\u001b[0m\n\u001b[1;32m   1421\u001b[0m                     \u001b[0;31m#  we may try to only slice\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1422\u001b[0m                     \u001b[0mtaker\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mblklocs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmgr_locs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1423\u001b[0;31m                     \u001b[0mmax_len\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmax\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmgr_locs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtaker\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmax\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1424\u001b[0m                     \u001b[0;32mif\u001b[0m \u001b[0monly_slice\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1425\u001b[0m                         \u001b[0mtaker\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmaybe_indices_to_slice\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtaker\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmax_len\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/gxe2021/lib/python3.8/site-packages/numpy/core/_methods.py\u001b[0m in \u001b[0;36m_amax\u001b[0;34m(a, axis, out, keepdims, initial, where)\u001b[0m\n\u001b[1;32m     37\u001b[0m def _amax(a, axis=None, out=None, keepdims=False,\n\u001b[1;32m     38\u001b[0m           initial=_NoValue, where=True):\n\u001b[0;32m---> 39\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mumr_maximum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkeepdims\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minitial\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mwhere\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     40\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     41\u001b[0m def _amin(a, axis=None, out=None, keepdims=False,\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "from itertools import product\n",
    "from rich.progress import track\n",
    "from scipy.stats import pearsonr\n",
    "\n",
    "mets = metabolites_and_microbes.columns[metabolites_and_microbes.columns.str.startswith(\"metabolite\")]\n",
    "mics = genera_wide_clr.columns\n",
    "\n",
    "combinations = list(product(mets, mics))\n",
    "\n",
    "metrics = pd.DataFrame(columns=[\"r\", \"p\", \"n\"], dtype=\"float64\", index=combinations)\n",
    "formula = \"{} ~ Q('{}')\"\n",
    "for comb in track(combinations):\n",
    "    data = metabolites_and_microbes[list(comb)].dropna()\n",
    "    ptest = pearsonr(data.iloc[:, 0], data.iloc[:, 1])\n",
    "    metrics.loc[[comb], \"r\"] = ptest[0]\n",
    "    metrics.loc[[comb], \"p\"] = ptest[1]\n",
    "    metrics.loc[[comb], \"n\"] = data.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "southern-evolution",
   "metadata": {},
   "outputs": [],
   "source": [
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "\n",
    "metrics[\"q\"] = fdrcorrection(metrics.p)[1]\n",
    "metrics[\"r2\"] = metrics.r.pow(2)\n",
    "metrics[\"metabolite\"] = metrics.index.map(lambda x: x[0])\n",
    "metrics[\"taxon\"] = metrics.index.map(lambda x: x[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "incoming-washer",
   "metadata": {},
   "outputs": [],
   "source": [
    "metrics.sort_values(by=\"p\").head(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "unsigned-filling",
   "metadata": {},
   "outputs": [],
   "source": [
    "metrics[metrics.q < 0.1].to_csv(\"data/sig_metabolite_taxon.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "million-copyright",
   "metadata": {},
   "outputs": [],
   "source": [
    "metrics[metrics.q < 0.05].metabolite.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "clear-investigator",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
